### ISSUE TITLE
EI_EXPOSE_REP2 being reported based off purely on method name

### ISSUE THREAD

**Issue by pnsantos** [spotbugs/spotbugs#1797]
Not really sure if this is a bug or a feature, but it seems spotbugs (4.3+) is reporting `EI_EXPOSE_REP2` purely based off the method name. For example I have the following:
```java
public class SomeClass {

  public CompletableFuture<Void> delete(String id) {
    // delete something in the db
  }

}
```
and then
```java
public class AnotherClass {

  private final SomeClass repo;

  public AnotherClass(SomeClass repo) {
    this.repo = repo;
  }

}
```

- Spotbugs 4.2 reports no violations
- Spotbugs 4.3+ reports `EI_EXPOSE_REP2` (tested all the way to 4.4.2)

However if I rename `delete` to something else (say `deleete`) then it no longer reports `EI_EXPOSE_REP2` which makes me think it's just looking at the method name?

Kinda makes this Bug useless... All our projects are suddenly reporting hundreds of `EI_EXPOSE_REP2`

Is this expected behaviour?

**Comment by pensionarchitects-ben** [comment]
We also have a problem with this, we upgraded to spotbugs 4.5.3 (from 4.0.0) and suddenly hundreds of errors while most of them are false positives. we disabled the warning for now. Subscribed to hear about progress on this.

**Comment by bmuniz-daitan** [comment]
Same here. Lots of reports of `EI_EXPOSE_REP2` with the new versions of `spotbugs`, some might be false positives. Also subscribing to hear about progress on this.

**Comment by gtoison** [comment]
`MutableClasses.looksLikeASetter` is used to detect a mutable class, it checks if a method name starts with one of the suspicious prefixes and if the return type is different compared to the parameter type:

https://github.com/spotbugs/spotbugs/blob/2fdbb01514683c32c07aeb841879269dddccbafc/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MutableClasses.java#L103-L108

https://github.com/spotbugs/spotbugs/blob/2fdbb01514683c32c07aeb841879269dddccbafc/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MutableClasses.java#L62-L64

I think the heuristic is a bit too aggressive here: considering the interest in this issue, there are many false positives detected.
@baloghadamsoftware what do you think?

**Comment by hazendaz** [comment]
fixed via #2514

**Comment by gtoison** [comment]
Hello @hazendaz, the issue is not fixed unfortunately: #2514 mitigates the problem for enums and records, but other classes might be recognized as mutable solely because they have a method whose name starts with one of the suspcious prefixes:

https://github.com/spotbugs/spotbugs/blob/2fdbb01514683c32c07aeb841879269dddccbafc/spotbugs/src/main/java/edu/umd/cs/findbugs/util/MutableClasses.java#L62-L64

One way to improve would be to let the users supply their own "setter like prefixes"?
